GeographicLib-2.1/man/GeodesicProj.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.43)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GEODESICPROJ 1"
.TH GEODESICPROJ 1 "2022-05-06" "GeographicLib 2.1" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
GeodesicProj \-\- perform projections based on geodesics
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBGeodesicProj\fR ( \fB\-z\fR | \fB\-c\fR | \fB\-g\fR ) \fIlat0\fR \fIlon0\fR [ \fB\-r\fR ]
[ \fB\-e\fR \fIa\fR \fIf\fR ] [ \fB\-w\fR ] [ \fB\-p\fR \fIprec\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Perform projections based on geodesics.  Convert geodetic coordinates to
either azimuthal equidistant, Cassini-Soldner, or gnomonic coordinates.
The center of the projection (\fIlat0\fR, \fIlon0\fR) is specified by either
the \fB\-c\fR option (for Cassini-Soldner), the \fB\-z\fR option (for azimuthal
equidistant), or the \fB\-g\fR option (for gnomonic).  At least one of these
options must be given (the last one given is used).
.PP
Geodetic coordinates are provided on standard input as a set of lines
containing (blank separated) \fIlatitude\fR and \fIlongitude\fR (decimal
degrees or degrees, minutes, seconds); for details on the allowed
formats for latitude and longitude, see the \f(CW\*(C`GEOGRAPHIC COORDINATES\*(C'\fR
section of \fBGeoConvert\fR\|(1).  For each set of geodetic coordinates, the
corresponding projected coordinates \fIx\fR, \fIy\fR (meters) are printed on
standard output together with the azimuth \fIazi\fR (degrees) and
reciprocal scale \fIrk\fR.  For Cassini-Soldner, \fIazi\fR is the bearing of
the easting direction and the scale in the easting direction is 1 and
the scale in the northing direction is 1/\fIrk\fR.  For azimuthal
equidistant and gnomonic, \fIazi\fR is the bearing of the radial direction
and the scale in the azimuthal direction is 1/\fIrk\fR.  For azimuthal
equidistant and gnomonic, the scales in the radial direction are 1 and
1/\fIrk\fR^2, respectively.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-z\fR \fIlat0\fR \fIlon0\fR" 4
.IX Item "-z lat0 lon0"
use the azimuthal equidistant projection centered at latitude = \fIlat0\fR,
longitude = \fIlon0\fR.  The \fB\-w\fR flag can be used to swap the default
order of the 2 coordinates, provided that it appears before \fB\-z\fR.
.IP "\fB\-c\fR \fIlat0\fR \fIlon0\fR" 4
.IX Item "-c lat0 lon0"
use the Cassini-Soldner projection centered at latitude = \fIlat0\fR,
longitude = \fIlon0\fR.  The \fB\-w\fR flag can be used to swap the default
order of the 2 coordinates, provided that it appears before \fB\-c\fR.
.IP "\fB\-g\fR \fIlat0\fR \fIlon0\fR" 4
.IX Item "-g lat0 lon0"
use the ellipsoidal gnomonic projection centered at latitude = \fIlat0\fR,
longitude = \fIlon0\fR.  The \fB\-w\fR flag can be used to swap the default
order of the 2 coordinates, provided that it appears before \fB\-g\fR.
.IP "\fB\-r\fR" 4
.IX Item "-r"
perform the reverse projection.  \fIx\fR and \fIy\fR are given on standard
input and each line of standard output gives \fIlatitude\fR, \fIlongitude\fR,
\&\fIazi\fR, and \fIrk\fR.
.IP "\fB\-e\fR \fIa\fR \fIf\fR" 4
.IX Item "-e a f"
specify the ellipsoid via the equatorial radius, \fIa\fR and
the flattening, \fIf\fR.  Setting \fIf\fR = 0 results in a sphere.  Specify
\&\fIf\fR < 0 for a prolate ellipsoid.  A simple fraction, e.g., 1/297,
is allowed for \fIf\fR.  By default, the \s-1WGS84\s0 ellipsoid is used, \fIa\fR =
6378137 m, \fIf\fR = 1/298.257223563.
.IP "\fB\-w\fR" 4
.IX Item "-w"
toggle the longitude first flag (it starts off); if the flag is on, then
on input and output, longitude precedes latitude (except that, on input,
this can be overridden by a hemisphere designator, \fIN\fR, \fIS\fR, \fIE\fR,
\&\fIW\fR).
.IP "\fB\-p\fR \fIprec\fR" 4
.IX Item "-p prec"
set the output precision to \fIprec\fR (default 6).  \fIprec\fR is the number
of digits after the decimal point for lengths (in meters).  For
latitudes, longitudes, and azimuths (in degrees), the number of digits
after the decimal point is \fIprec\fR + 5.  For the scale, the number of
digits after the decimal point is \fIprec\fR + 6.
.IP "\fB\-\-comment\-delimiter\fR \fIcommentdelim\fR" 4
.IX Item "--comment-delimiter commentdelim"
set the comment delimiter to \fIcommentdelim\fR (e.g., \*(L"#\*(R" or \*(L"//\*(R").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing and subsequently appended to the output line (separated by a
space).
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
print version and exit.
.IP "\fB\-h\fR" 4
.IX Item "-h"
print usage and exit.
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR \fIinfile\fR" 4
.IX Item "--input-file infile"
read input from the file \fIinfile\fR instead of from standard input; a file
name of \*(L"\-\*(R" stands for standard input.
.IP "\fB\-\-input\-string\fR \fIinstring\fR" 4
.IX Item "--input-string instring"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR \fIlinesep\fR" 4
.IX Item "--line-separator linesep"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR \fIoutfile\fR" 4
.IX Item "--output-file outfile"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of \*(L"\-\*(R" stands for standard output.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.Vb 4
\&   echo 48.648 \-2.007 | GeodesicProj \-c 48.836 2.337
\&   => \-319919 \-11791 86.7 0.999
\&   echo \-319919 \-11791 | GeodesicProj \-c 48.836 2.337 \-r
\&   => 48.648 \-2.007 86.7 0.999
.Ve
.SH "ERRORS"
.IX Header "ERRORS"
An illegal line of input will print an error message to standard output
beginning with \f(CW\*(C`ERROR:\*(C'\fR and causes \fBGeodesicProj\fR to return an exit
code of 1.  However, an error does not cause \fBGeodesicProj\fR to
terminate; following lines will be converted.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
The ellipsoidal gnomonic projection is derived in Section 8 of
C. F. F. Karney, \fIAlgorithms for geodesics\fR, J. Geodesy 87, 43\-55 (2013); \s-1DOI\s0
<https://doi.org/10.1007/s00190\-012\-0578\-z>;
addenda: <https://geographiclib.sourceforge.io/geod\-addenda.html>.
.SH "AUTHOR"
.IX Header "AUTHOR"
\&\fBGeodesicProj\fR was written by Charles Karney.
.SH "HISTORY"
.IX Header "HISTORY"
\&\fBGeodesicProj\fR was added to GeographicLib,
<https://geographiclib.sourceforge.io>, in 2009\-08.
Prior to version 1.9 it was called EquidistantTest.


ros_env/history.txt
 1071  rostopic list
 1072  sshm
 1073  roslaunch omo_r1mini_bringup omo_r1_imu.launch
 1074  cm
 1075  roslaunch omo_r1mini_bringup omo_r1_imu.launch
 1076  roslaunch omo_r1_teleop omo_r1_teleop_key.launch 
 1077  roscore
 1078  nb
 1079  eb
 1080  roscore
 1081  sshr
 1082  sshr
 1083  anydesk
 1084  roscore
 1085  sshm
 1086  sudo apt install ros-melodic-cartographer-ros
 1087  roscore
 1088  gazebo
 1089  rviz
 1090  sshm
 1091  roscore
 1092  roslaunch omo_r1mini_cartographer omo_r1mini_cartographer_rviz.launch
 1093  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launc
 1094  rosclean
 1095  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launc
 1096  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1097  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1098  cd {$HOME}
 1099  hp,e
 1100  cd home
 1101  sshm
 1102  sshm
 1103  cd to map folder
 1104  cs
 1105  cd ..
 1106  ls
 1107  cd map
 1108  rosrun map_server map_saver -f adyslam
 1109  roscore
 1110  roslaunch omo_r1mini_cartographer omo_r1mini_cartographer_rviz.launch
 1111  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1112  roscore
 1113  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1114  sshm
 1115  sshm
 1116  roscore
 1117  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1118  roslaunch omo_r1mini_navigation omo_r1mini_navigation.launch
 1119  rviz
 1120  roscore
 1121  rviz
 1122  roscore
 1123  sshm
 1124  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1125  sshm
 1126  roscore
 1127  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1128  sshm1
 1129  sshm
 1130  roscore
 1131  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1132  sudo systemctl restart t
 1133  sshm
 1134  sshm
 1135  roscore
 1136  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1137  roscore
 1138  sshm
 1139  rosrun map_server map_server map_file.yaml
 1140  [출처] ROS Navigation Map Creation|작성자 Tomboy
 1141  rosrun map_server map_server map.yaml
 1142  rosrun map_server map_server adyslam.yaml
 1143  roscore
 1144  cs
 1145  cd omo_r1mini/omo_r1mini_navigation/maps
 1146  rosrun map_server map_saver -f adyslam2
 1147  ./catkin_ws/src/omo_r1mini/omo_r1mini_cartographer/scripts/save_pbstream.sh
 1148  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1149  roslaunch omo_r1mini_cartographer omo_r1mini_cartographer_rviz.launch
 1150  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1151  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1152  sshm
 1153  roscore
 1154  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1155  roscore
 1156  rosrun map_server map_saver -f map.yaml
 1157  rosrun map_server map_saver -f map
 1158  roscore
 1159  sshm
 1160  roscore
 1161  roslaunch omo_r1mini_cartographer omo_r1mini_cartographer_rviz.launch
 1162  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launc
 1163  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1164  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1165  cs
 1166  cd omo_r1mini/omo_r1mini_navigation/maps
 1167  rosrun map_server map_saver -f tutorial
 1168  roscore
 1169  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1170  rqt_graph
 1171  sshm
 1172  sshm
 1173  sshm
 1174  roscore
 1175  sshm
 1176  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1177  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1178  sshm
 1179  sshm
 1180  sshm
 1181  sshm
 1182  roscore
 1183  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1184  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1185  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1186  roscore
 1187  sshm
 1188  sshm
 1189  roscore
 1190  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1191  roslaunch omo_r1mini_teleop omo_r1mini_teleop_key.launch
 1192  roscore
 1193  rosrun map_server map_server map.yaml
 1194  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1195  roscore
 1196  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1197  rosrun map_server map_server map.yaml
 1198  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1199  cp map.*home/
 1200  sshm
 1201  sshm
 1202  roscore
 1203  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1204  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1205  sshm
 1206  sshm
 1207  roscore
 1208  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1209  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1210  sshm
 1211  roscore
 1212  sshm
 1213  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1214  roslaunch omo_r1mini_slam omo_r1mini_slam_rviz.launch
 1215  sshm
 1216  sshm
 1217  ls
 1218  sshm
 1219  roscore
 1220  sshm 
 1221  sshm
 1222  gogo
 1223  sshm
 1224  roscore
 1225  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1226  sshm
 1227  sshm
 1228  roscore
 1229  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1230  sshm
 1231  roscore
 1232  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1233  sshm
 1234  roscore
 1235  sshm1
 1236  sshm
 1237  roscore
 1238  roslaunch omo_r1mini_navigation omo_r1mini_navigation_rviz.launch
 1239  sshm
 1240  sshm
 1241  git clone nav2_bt_navigator/behavior_trees/navigate_through_poses_w_replanning_and_recovery.xml
 1242  sudo apt-get install git-core
 1243  sudo apt-get install update
 1244  git --version
 1245  sudo git config --global user.name "ahndy"
 1246  sudo git config --global user.email "ahndy@mobilio.io"
 1247  sudo git config --global color.ui "auto"
 1248  mkdir mygit
 1249  cd mygit
 1250  ls
 1251  cs
 1252  cd
 1253  sudo git init
 1254  git status
 1255  cd mygit
 1256  git init
 1257  git clone https://github.com/Ahndy123/navigation2.git
 1258  git clone 

GeographicLib-1.34/tools/GeodesicProj.cpp
/**
 * \file GeodesicProj.cpp
 * \brief Command line utility for geodesic projections
 *
 * Copyright (c) Charles Karney (2009-2012) <charles@karney.com> and licensed
 * under the MIT/X11 License.  For more information, see
 * http://geographiclib.sourceforge.net/
 *
 * Compile and link with
 *   g++ -g -O3 -I../include -I../man -o GeodesicProj \
 *       GeodesicProj.cpp \
 *       ../src/AzimuthalEquidistant.cpp \
 *       ../src/CassiniSoldner.cpp \
 *       ../src/DMS.cpp \
 *       ../src/Geodesic.cpp \
 *       ../src/GeodesicLine.cpp \
 *       ../src/Gnomonic.cpp
 *
 * See the <a href="GeodesicProj.1.html">man page</a> for usage
 * information.
 **********************************************************************/

#include <iostream>
#include <sstream>
#include <string>
#include <sstream>
#include <fstream>
#include <GeographicLib/Geodesic.hpp>
#include <GeographicLib/AzimuthalEquidistant.hpp>
#include <GeographicLib/CassiniSoldner.hpp>
#include <GeographicLib/Gnomonic.hpp>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>

#if defined(_MSC_VER)
// Squelch warnings about constant conditional expressions and potentially
// uninitialized local variables
#  pragma warning (disable: 4127 4701)
#endif

#include "GeodesicProj.usage"

int main(int argc, char* argv[]) {
  try {
    using namespace GeographicLib;
    typedef Math::real real;
    bool azimuthal = false, cassini = false, gnomonic = false, reverse = false;
    real lat0 = 0, lon0 = 0;
    real
      a = Constants::WGS84_a<real>(),
      f = Constants::WGS84_f<real>();
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';';

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-r")
        reverse = true;
      else if (arg == "-c" || arg == "-z" || arg == "-g") {
        cassini = arg == "-c";
        azimuthal = arg == "-z";
        gnomonic = arg == "-g";
        if (m + 2 >= argc) return usage(1, true);
        try {
          DMS::DecodeLatLon(std::string(argv[m + 1]), std::string(argv[m + 2]),
                            lat0, lon0);
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of " << arg << ": "
                    << e.what() << "\n";
          return 1;
        }
        m += 2;
      } else if (arg == "-e") {
        if (m + 2 >= argc) return usage(1, true);
        try {
          a = Utility::num<real>(std::string(argv[m + 1]));
          f = Utility::fract<real>(std::string(argv[m + 2]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -e: " << e.what() << "\n";
          return 1;
        }
        m += 2;
      } else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout
          << argv[0] << ": GeographicLib version "
          << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else
        return usage(!(arg == "-h" || arg == "--help"), arg != "--help");
    }

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    if (!(azimuthal || cassini || gnomonic)) {
      std::cerr << "Must specify \"-z lat0 lon0\" or "
                << "\"-c lat0 lon0\" or \"-g lat0 lon0\"\n";
      return 1;
    }

    const Geodesic geod(a, f);
    const CassiniSoldner cs = cassini ?
      CassiniSoldner(lat0, lon0, geod) : CassiniSoldner(geod);
    const AzimuthalEquidistant az(geod);
    const Gnomonic gn(geod);

    std::string s;
    int retval = 0;
    std::cout << std::fixed;
    while (std::getline(*input, s)) {
      try {
        std::string eol("\n");
        if (!cdelim.empty()) {
          std::string::size_type m = s.find(cdelim);
          if (m != std::string::npos) {
            eol = " " + s.substr(m) + "\n";
            s = s.substr(0, m);
          }
        }
        std::istringstream str(s);
        real lat, lon, x, y, azi, rk;
        std::string stra, strb;
        if (!(str >> stra >> strb))
          throw GeographicErr("Incomplete input: " + s);
        if (reverse) {
          x = Utility::num<real>(stra);
          y = Utility::num<real>(strb);
        } else
          DMS::DecodeLatLon(stra, strb, lat, lon);
        std::string strc;
        if (str >> strc)
          throw GeographicErr("Extraneous input: " + strc);
        if (reverse) {
          if (cassini)
            cs.Reverse(x, y, lat, lon, azi, rk);
          else if (azimuthal)
            az.Reverse(lat0, lon0, x, y, lat, lon, azi, rk);
          else
            gn.Reverse(lat0, lon0, x, y, lat, lon, azi, rk);
          *output << Utility::str<real>(lat, 15) << " "
                  << Utility::str<real>(lon, 15) << " "
                  << Utility::str<real>(azi, 15) << " "
                  << Utility::str<real>(rk, 16) << eol;
        } else {
          if (cassini)
            cs.Forward(lat, lon, x, y, azi, rk);
          else if (azimuthal)
            az.Forward(lat0, lon0, lat, lon, x, y, azi, rk);
          else
            gn.Forward(lat0, lon0, lat, lon, x, y, azi, rk);
          *output << Utility::str<real>(x, 10) << " "
                  << Utility::str<real>(y, 10) << " "
                  << Utility::str<real>(azi, 15) << " "
                  << Utility::str<real>(rk, 16) << eol;
        }
      }
      catch (const std::exception& e) {
        *output << "ERROR: " << e.what() << "\n";
        retval = 1;
      }
    }
    return retval;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
